File: src/libflatsurf/flat_triangulation.cc
/**********************************************************************
 *  This file is part of flatsurf.
 *
 *        Copyright (C) 2019 Julian Rüth
 *
 *  Flatsurf is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  Flatsurf is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with flatsurf. If not, see <https://www.gnu.org/licenses/>.
 *********************************************************************/

#include <ostream>
#include <vector>

#include "flatsurf/assert.hpp"
#include "flatsurf/flat_triangulation.hpp"
#include "flatsurf/half_edge.hpp"
#include "flatsurf/half_edge_map.hpp"

using std::ostream;
using std::vector;

namespace flatsurf {
namespace {
template <typename T>
void updateAfterFlip(HalfEdgeMap<T> &map, HalfEdge halfEdge,
                     const FlatTriangulationCombinatorial &parent) {
  map.set(halfEdge, map.get(-parent.nextInFace(halfEdge)) +
                        map.get(parent.nextAtVertex(halfEdge)));
}
}  // namespace

template <typename Vector>
class FlatTriangulation<Vector>::Implementation {
 public:
  Implementation(HalfEdgeMap<Vector> &&vectors) : vectors(std::move(vectors)) {}

  const HalfEdgeMap<Vector> vectors;
};

template <typename Vector>
const Vector &FlatTriangulation<Vector>::fromEdge(const HalfEdge e) const {
  return impl->vectors.get(e);
}

template <typename Vector>
FlatTriangulation<Vector>::FlatTriangulation(
    FlatTriangulationCombinatorial &&combinatorial,
    const vector<Vector> &vectors)
    : FlatTriangulation(std::move(combinatorial),
                        HalfEdgeMap<Vector>(combinatorial, vectors,
                                            updateAfterFlip<Vector>)) {}

template <typename Vector>
FlatTriangulation<Vector>::FlatTriangulation(
    FlatTriangulationCombinatorial &&combinatorial,
    HalfEdgeMap<Vector> &&vectors)
    : FlatTriangulationCombinatorial(std::move(combinatorial)),
      impl(spimpl::make_unique_impl<Implementation>(std::move(vectors))) {
  // check that faces are closed
  for (auto edge : halfEdges()) {
    auto zero = fromEdge(edge);
    edge = nextInFace(edge);
    zero += fromEdge(edge);
    edge = nextInFace(edge);
    zero += fromEdge(edge);
    CHECK_ARGUMENT(!zero, "some face is not closed");
  }
}

template <typename Vector>
FlatTriangulation<Vector> FlatTriangulation<Vector>::clone() const {
  std::vector<Vector> vectors;
  for (int e = 1; e <= halfEdges().size() / 2; e++)
    vectors.push_back(fromEdge(HalfEdge(e)));
  return FlatTriangulation(FlatTriangulationCombinatorial::clone(),
                           std::move(vectors));
}

template <typename Vector>
ostream &operator<<(ostream &os, const FlatTriangulation<Vector> &self) {
  return os << static_cast<const FlatTriangulationCombinatorial &>(self)
            << " with vectors " << self.impl->vectors;
}
}  // namespace flatsurf

// Instantiations of templates so implementations are generated for the linker
#include <exact-real/number_field_traits.hpp>
#include "flatsurf/vector_eantic.hpp"
#include "flatsurf/vector_exactreal.hpp"
#include "flatsurf/vector_longlong.hpp"

using namespace flatsurf;

template class flatsurf::FlatTriangulation<VectorLongLong>;
template ostream &flatsurf::operator<<(ostream &, const FlatTriangulation<VectorLongLong> &);
template class flatsurf::FlatTriangulation<VectorEAntic>;
template ostream &flatsurf::operator<<(ostream &, const FlatTriangulation<VectorEAntic> &);
template class flatsurf::FlatTriangulation<VectorExactReal<exactreal::NumberFieldTraits>>;
template ostream &flatsurf::operator<<(ostream &, const FlatTriangulation<VectorExactReal<exactreal::NumberFieldTraits>> &);

File: pyflatsurf/test/flat_triangulation.py
#!/usr/bin/env python

######################################################################
# This file is part of flatsurf.
#
#       Copyright (C) 2020-2025 Julian Rüth
#
# flatsurf is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# flatsurf is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with flatsurf. If not, see <https://www.gnu.org/licenses/>.
######################################################################

import sys
import pytest
import cppyy

from pyflatsurf import flatsurf
import surfaces

def test_addition():
    vector = flatsurf.Vector['mpq_class']

    L = surfaces.L(vector)

    # A stretched L
    shift = [vector(0, L.fromHalfEdge(e.positive()).y()) for e in L.edges()]
    L = (L + shift).surface()

    for halfEdge in L.halfEdges():
        assert L.fromHalfEdge(halfEdge).y() != 1

def test_isomorphism():
    vector = flatsurf.Vector['mpq_class']
    L = surfaces.L(vector)
    assert L.isomorphism(L)
    assert L.isomorphism(L, filter_matrix=lambda a, b, c, d: a == -1 and b == 0 and c == 0 and d == -1)
    assert not L.isomorphism(L, filter_matrix=lambda a, b, c, d: a*d - b*c not in [-1, 1])

def test_serialization():
    hexagon = surfaces.random_hexagon(flatsurf.Vector['exactreal::Element<exactreal::NumberField>'])

    from pickle import loads, dumps
    clone = loads(dumps(hexagon))
    assert clone == hexagon
    assert hash(clone) == hash(hexagon)

def test_readme():
    # Make sure to update the README.md if you make any changes to this test case.
    import pyflatsurf
    import cppyy
    import pyeantic

    K = pyeantic.eantic.renf_class.make("x^2 - 3", "x", "1.73 +/- 0.1")
    x = K.gen()
    R2 = pyflatsurf.flatsurf.Vector[cppyy.gbl.eantic.renf_elem_class]
    R = pyeantic.eantic.renf_elem
    vectors = [R2(R(K, 2), R(K, 0)), R2(R(K, 1), x), R2(R(K, 3), x), R2(R(K, 1), -x), R2(R(K, 4), R(K, 0)), R2(R(K, 3), x)]
    vertices = [[1, 3, -4, -5, -3, -2], [2, -1, -6, 4, 5, 6]]
    S = pyflatsurf.Surface(vertices, vectors)

    C = S.connections().bound(4)
    assert "\n".join(str(c) for c in C) == "1\n-1\n2\n(-3, (x ~ 1.7320508)) from 2 to 4\n(0, (2*x ~ 3.4641016)) from 2 to -6\n(-2, (2*x ~ 3.4641016)) from 2 to -2\n-2\n(3, (-x ~ -1.7320508)) from -2 to -4\n(0, (-2*x ~ -3.4641016)) from -2 to 3\n(2, (-2*x ~ -3.4641016)) from -2 to 2\n3\n(2, (2*x ~ 3.4641016)) from 3 to -6\n(0, (2*x ~ 3.4641016)) from 3 to -2\n-3\n4\n(3, (-x ~ -1.7320508)) from 4 to 2\n-4\n(-3, (x ~ 1.7320508)) from -4 to -2\n5\n-5\n6\n-6\n(-2, (-2*x ~ -3.4641016)) from -6 to 3\n(0, (-2*x ~ -3.4641016)) from -6 to 2"


if __name__ == '__main__': sys.exit(pytest.main(sys.argv))
